crates/sui-framework/src/natives/crypto/hash.rs

```diff
@@ -1,50 +1,116 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
-use crate::legacy_empty_cost;
+use crate::natives::NativesCostTable;
 use fastcrypto::hash::{Blake2b256, HashFunction, Keccak256};
 use move_binary_format::errors::PartialVMResult;
-use move_vm_runtime::native_functions::NativeContext;
+use move_core_types::gas_algebra::InternalGas;
+use move_vm_runtime::{native_charge_gas_early_exit, native_functions::NativeContext};
 use move_vm_types::{
     loaded_data::runtime_types::Type,
     natives::function::NativeResult,
     pop_arg,
     values::{Value, VectorRef},
 };
 use smallvec::smallvec;
-use std::collections::VecDeque;
+use std::{collections::VecDeque, ops::Mul};
 
 fn hash<H: HashFunction<DIGEST_SIZE>, const DIGEST_SIZE: usize>(
-    _context: &mut NativeContext,
+    context: &mut NativeContext,
     ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
+    // The caller provides the cost per byte
+    msg_cost_per_byte: InternalGas,
 ) -> PartialVMResult<NativeResult> {
     debug_assert!(ty_args.is_empty());
     debug_assert!(args.len() == 1);
 
-    // TODO: implement native gas cost estimation https://github.com/MystenLabs/sui/issues/3593
-    let cost = legacy_empty_cost();
     let msg = pop_arg!(args, VectorRef);
+    let msg_ref = msg.as_bytes_ref();
+
+    // Charge the msg dependent costs
+    native_charge_gas_early_exit!(
+        context,
+        msg_cost_per_byte.mul((msg_ref.len() as u64).into())
+    );
 
     Ok(NativeResult::ok(
-        cost,
+        context.gas_used(),
         smallvec![Value::vector_u8(
             H::digest(msg.as_bytes_ref().as_slice()).digest
         )],
     ))
 }
 
+#[derive(Clone)]
+pub struct HashKeccak256CostParams {
+    /// Base cost for invoking the `blake2b256` function
+    pub hash_keccak256_cost_base: InternalGas,
+    /// Cost per byte of `data`
+    pub hash_keccak256_data_cost_per_byte: InternalGas,
```

Comment:
Hashing 2 bytes or 3 costs the same - Should we change this to cost_per_block? or do we include in it also the move to Rust deserialization and thus we better stay with message length?

Reply:
I'm open to either as it's maintainable and best represent how the cost scales with input length.

Reply:
Yes, Keccak256 has a block size of 32 bytes, so maybe have here:
- hash_keccak256_data_cost_per_byte - to represent the cost of deserialization between Move and Rust
- hash_keccak256_data_cost_per_block - to represent the actual hash operation (using msg.len()/32*hash_keccak256_data_cost_per_block)

For blake2b it is 128 bytes.

Reply:
So for blake2b it'll be` msg.len()/128 * data_cost_per_block` right?

Reply:
Yes.

@kchalkias wdyt?

Reply:
yes that's correct, supporting this method to use block size for hash functions Vs single bytes

Reply:
Yes that's correct, supporting this method to use block size for hash functions Vs single bytes

Reply:
Re 
> hash_keccak256_data_cost_per_byte - to represent the cost of deserialization between Move and Rust
I'm not sure for that, cause we have to take it into consideration everywhere probably (on native functions) we have to test if hashing operation per block is many many times more expensive than deser cost.

Reply:
fixed